core/matrix/ell.hpp

```diff
@@ -0,0 +1,271 @@
+     */
+    value_type &val_at(size_type row, size_type col) noexcept
+    {
+        return values_.get_data()[linearize_index(row, col)];
+    }
+
+    /**
+     * @copydoc Ell::val_at(size_type, size_type)
+     */
+    value_type val_at(size_type row, size_type col) const noexcept
+    {
+        return values_.get_const_data()[linearize_index(row, col)];
+    }
+
+    /**
+     * Returns the `idx`-th column index of the `row`-th row .
+     *
+     * @param row  the row of the requested element
+     * @param col  the column of the requested element
+     *
+     * @note  the method has to be called on the same Executor the matrix is
+     *        stored at (e.g. trying to call this method on a GPU matrix from
+     *        the CPU results in a runtime error)
+     */
+    index_type &col_at(size_type row, size_type col) noexcept
+    {
+        return get_col_idxs()[linearize_index(row, col)];
+    }
+
+    /**
+     * @copydoc Ell::col_at(size_type, size_type)
+     */
+    index_type col_at(size_type row, size_type col) const noexcept
+    {
+        return get_const_col_idxs()[linearize_index(row, col)];
+    }
+
+protected:
+    /**
+     * Creates an empty ELL matrix.
+     *
+     * @param exec  Executor associated to the matrix
+     */
+    explicit Ell(std::shared_ptr<const Executor> exec)
+        : BasicLinOp<Ell>(exec, 0, 0, 0),
+          values_(exec),
+          col_idxs_(exec),
+          max_nonzeros_per_row_(0),
+          padding_(0)
+    {}
+
+    /**
+     * Creates an uninitialized Ell matrix of the specified size.
+     *
+     * @param exec  Executor associated to the matrix
+     * @param num_rows               number of rows
+     * @param num_cols               number of columns
+     * @param max_nonzeros_per_row   maximum number of nonzeros in one row
+     * @param padding                padding of the rows
+     */
+    Ell(std::shared_ptr<const Executor> exec, size_type num_rows,
+        size_type num_cols, size_type max_nonzeros_per_row, size_type padding)
+        : BasicLinOp<Ell>(exec, num_rows, num_cols,
+                          padding*max_nonzeros_per_row),
+          values_(exec, padding*max_nonzeros_per_row),
+          col_idxs_(exec, padding*max_nonzeros_per_row),
+          max_nonzeros_per_row_(max_nonzeros_per_row),
+          padding_(padding)
+    {}
+
+    /**
+     * Creates an uninitialized Ell matrix of the specified size.
+     *    (The padding is set to the number of rows of the matrix.)
+     *
+     * @param exec  Executor associated to the matrix
+     * @param num_rows               number of rows
+     * @param num_cols               number of columns
+     * @param max_nonzeros_per_row   maximum number of nonzeros in one row
+     */
+    Ell(std::shared_ptr<const Executor> exec, size_type num_rows,
+        size_type num_cols, size_type max_nonzeros_per_row)
+        : Ell(std::move(exec), num_rows, num_cols,
+              max_nonzeros_per_row, num_rows)
```

Comment:
Add to be consistent:
```c++
padding_(num_rows)
```

In general, never leave primitive types uninitialized in an object, because then the object is in a corrupted state.

Reply:
Ah, sorry about the comment, I didn't realize this was calling another Ell constructor. Though it was BasicLinOp<>.